daemon/daemon_main.go

```diff
@@ -1328,6 +1335,37 @@ func runDaemon() {
 		}()
 	}
 
+	// Migrating the ENI datapath must happen before the API is served to
+	// prevent endpoints from being created. It also must be before the health
+	// initialization logic which creates the health endpoint, for the same
+	// reasons as the API being served. We want to ensure that this migration
+	// logic runs before any endpoint creates.
+	if option.Config.IPAM == option.IPAMENI {
+		migrated, partial, err, revErr := linuxrouting.NewMigrator(
+			&interfaceNumberGetter{},
+		).MigrateENIDatapath(option.Config.EgressMultiHomeIPRuleCompat)
+		switch {
+		case err != nil && revErr != nil:
+			log.WithError(err).Fatal("Failed to migrate ENI datapath. The original datapath failed to be reinstated. " +
+				"Connectivity is not guaranteed and manual intervention is required.")
```

Comment:
We had some discussion yesterday about the `Fatal()` here, the concern being that at a high level if somehow migration fails and the rollback also fails, then the node will be left in a "hybrid" state with both new and old route rules/tables and the upgrade will fail with the Cilium pod exiting out fatally due to this line.

Then when downgrading, if we also attempt to migrate back to the old scheme and that fails, in the previous version that would also try to rollback whatever changes were made while trying to migrate back to the old scheme, and then that deployment would also trigger the `Fatal()` here.

So in the worst case, the upgrade would fail with crashing Cilium, then the instructions tell the user to disable the compat flag and maybe that succeeds but worst case it also fails and Cilium crashes and then the user is probably left guessing exactly what to do but presumably would roll back to v1.7.12 or earlier.

I didn't yet review the most recent changes since last night, just wanted to check whether this particular concern was considered in the last changes.

Reply:
Just to sync here from our offline discussion. 

We discussed the need for `Fatal` and determined that it is not necessary to `Fatal`. This is because we can guarantee an atomic datapath migration. Due to the order in which we migrate to another datapath, an endpoint can either have the old datapath or the new datapath, at any point, regardless of failures. (It is possible to have both datapaths in-place as [described here](https://github.com/cilium/cilium/pull/14337#discussion_r546999794), but that doesn't cause any issues because the rule and routes only differ by table ID, but have the same routing info anyway).

Instead, we will warn the user using log msgs to indicate if anything had gone wrong and when a warning is serious enough to warrant trying the migration again. The worse case is when the user doesn't heed the warnings, leading to future conflicts with new endpoint. The warnings advise cleaning up certain state that was left over in case of a failure in cleaning up in the migration.